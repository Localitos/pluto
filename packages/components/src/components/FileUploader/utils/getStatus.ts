import isEmpty from "lodash/isEmpty";
import { FileUploaderStatus } from "../types/FileUploaderStatus";

export const getStatus = ({
  progress,
  fileUrl,
  errorMessage,
  required,
}: {
  progress: number;
  fileUrl?: string;
  errorMessage?: string;
  required?: boolean;
}): FileUploaderStatus => {
  if (progress > 0 && progress < 100) return "loading";
  if (!isEmpty(fileUrl)) return "success";
  if (!isEmpty(errorMessage)) return "error";
  if (!isEmpty(errorMessage) && required && !fileUrl) return "error";
  return "waiting";
};
