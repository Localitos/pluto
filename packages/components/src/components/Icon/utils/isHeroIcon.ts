import * as HeroOutlineIcons from "../HeroIcons";
import { IconName } from "../types/IconName";

export const isHeroIcon = (
  name: IconName,
): name is keyof typeof HeroOutlineIcons => {
  return name in HeroOutlineIcons;
};
