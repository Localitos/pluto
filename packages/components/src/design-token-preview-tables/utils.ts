import replace from "lodash/replace";
import map from "lodash/map";
import toUpper from "lodash/toUpper";
import split from "lodash/split";
import { ColorToken, GeneralToken } from "./types";

export const copyToClipboard =
  (textToCopy: string): (() => void) =>
  () => {
    // Create a temporary input element
    const input = document.createElement("input");
    input.value = textToCopy;

    // Append the input element to the DOM
    document.body.append(input);

    // Select the text in the input element
    input.select();

    // Copy the selected text to the clipboard
    document.execCommand("copy");

    // Remove the input element
    input.remove();
  };

// this function helps replicate how to use the space token in component props. e.g. <Button paddingBottom="space30" />
export const formatTokenName = (token: GeneralToken, prefix: string): string =>
  `${prefix}${
    toUpper(token[0].charAt(0)) + replace(token[0].slice(1), "-", "")
  }`;


