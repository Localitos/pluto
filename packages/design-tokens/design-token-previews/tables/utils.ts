import replace from "lodash/replace";
import map from "lodash/map";
import toUpper from "lodash/toUpper";
import split from "lodash/split";
import { ColorToken, GeneralToken } from "./types";

// this function helps replicate how to use the space token in component props. e.g. <Button paddingBottom="space30" />
export const formatTokenName = (token: GeneralToken, prefix: string): string =>
  `${prefix}${
    toUpper(token[0].charAt(0)) + replace(token[0].slice(1), "-", "")
  }`;

// this function helps replicate how to use the color token in component props. e.g. <Button color="colorTextError" />
export const formatColorTokenName = (token: ColorToken): string => {
  const prefix = "color";
  const words = split(token[0], "-");

  // Capitalize the first letter of each word and join them
  const formattedWords = map(
    words,
    (word) => toUpper(word.charAt(0)) + word.slice(1)
  ).join("");

  return prefix + formattedWords;
};

export const hexToRgb = (hex: string): string => {
  // Remove '#' if it's present
  hex = replace(hex, "#", "");

  // Parse the hex value into RGB components
  const r = Number.parseInt(hex.slice(0, 2), 16);
  const g = Number.parseInt(hex.slice(2, 4), 16);
  const b = Number.parseInt(hex.slice(4, 6), 16);

  return `rgb(${r}, ${g}, ${b})`;
};

export const hexToHsla = (hex: string, alpha = 1): string => {
  hex = replace(hex, "#", "");

  const r = Number.parseInt(hex.slice(0, 2), 16) / 255;
  const g = Number.parseInt(hex.slice(2, 4), 16) / 255;
  const b = Number.parseInt(hex.slice(4, 6), 16) / 255;

  const max = Math.max(r, g, b);
  const min = Math.min(r, g, b);
  const lightness = (max + min) / 2;

  let hue = 0,
    saturation: number;

  if (max === min) {
    hue = 0; // achromatic
    saturation = 0;
  } else {
    const delta = max - min;
    saturation =
      lightness > 0.5 ? delta / (2 - max - min) : delta / (max + min);
    switch (max) {
      case r: {
        hue = (g - b) / delta + (g < b ? 6 : 0);
        break;
      }
      case g: {
        hue = (b - r) / delta + 2;
        break;
      }
      case b: {
        hue = (r - g) / delta + 4;
        break;
      }
    }
    hue /= 6;
  }

  const h = Math.round(hue * 360);
  const s = Math.round(saturation * 100);
  const l = Math.round(lightness * 100);

  return `hsla(${h}, ${s}%, ${l}%, ${alpha})`;
};

export const formatHex = (token: ColorToken): string => token[1].value;

export const formatPixels = (token: GeneralToken): string | undefined =>
  token[1].comment;
export const formatRems = (token: GeneralToken): string => token[1].value;
