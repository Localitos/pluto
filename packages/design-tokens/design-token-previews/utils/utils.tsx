import React from "react";
import keys from "lodash/keys";
import camelCase from "lodash/camelCase";
import capitalize from "lodash/capitalize";

import { TokenEntry } from "../types/TokenEntry";

import { CopyToClipboardButton } from "../CopyToClipboardButton";

export const getKey = (object: Record<string, unknown>): string => {
  return keys(object)[0];
};

export const getValue = ([, token]: TokenEntry): string => {
  return token.value;
};

export const getComment = ([, token]: TokenEntry): string => {
  return token.comment || "";
};

export const formatTokenName = (
  prefix: string,
  suffix: string
): JSX.Element => {
  const tokenName = camelCase(`${prefix}${capitalize(suffix)}`);
  return (
    <div style={{ display: "flex" }}>
      <div>{tokenName}</div>
      <div>
        <CopyToClipboardButton textToCopy={tokenName} />
      </div>
    </div>
  );
};
