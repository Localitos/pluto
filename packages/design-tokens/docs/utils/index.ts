export * from "./getTokenKey";
export * from "./getTokenName";
export * from "./getTokenValue";
export * from "./getTokenComment";
export * from "./createFontSizePreview";
export * from "./colorConverters";
export * from "./buildTokensTableRows";
export * from "./copyToClipboard";
export * from "./getTokenNameFromTuple";
