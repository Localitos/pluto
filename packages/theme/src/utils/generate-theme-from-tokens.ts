/* eslint-disable eslint-comments/disable-enable-pair */
/* eslint-disable @typescript-eslint/no-explicit-any */
import type {
  GenericThemeShape,
  BorderWidthsKeys,
  ColorsKeys,
  FontSizesKeys,
  FontWeightsKeys,
  FontsKeys,
  LineHeightsKeys,
  RadiiKeys,
  BoxShadowsKeys,
  SpacingsKeys,
  ZIndicesKeys,
} from "../types/generic-theme-shape";

interface GenerateThemeFromTokensArgs {
  borderWidths: Partial<{ [key in BorderWidthsKeys]: any }>;
  colors: Partial<{ [key in ColorsKeys]: any }>;
  radii: Partial<{ [key in RadiiKeys]: any }>;
  fonts: Partial<{ [key in FontsKeys]: any }>;
  fontSizes: Partial<{ [key in FontSizesKeys]: any }>;
  fontWeights: Partial<{ [key in FontWeightsKeys]: any }>;
  lineHeights: Partial<{ [key in LineHeightsKeys]: any }>;
  boxShadows: Partial<{ [key in BoxShadowsKeys]: any }>;
  spacings: Partial<{ [key in SpacingsKeys]: any }>;
  zIndices: Partial<{ [key in ZIndicesKeys]: any }>;
}

/**
 * Generates an avatar background color using the length of a person's full name.
 *
 * @param theme - Object of individual theme objects.
 * @param theme.borderWidths - Border width values
 * @param theme.colors - Color values
 * @param theme.radii - Radius values
 * @param theme.fonts - Font family values
 * @param theme.fontSizes - Font size values
 * @param theme.fontWeights - Font weight values
 * @param theme.lineHeights - Line height values
 * @param theme.boxShadows - Box shadow values
 * @param theme.spacings - Spacing values
 * @param theme.zIndices - zIndex values
 * @returns GenericThemeShape.
 * @example
 * generateThemeFromTokens({
 * backgroundColors,
 * borderColors,
 * borderWidths,
 * colors,
 * radii,
 * fonts,
 * fontSizes,
 * fontWeights,
 * lineHeights,
 * boxShadows,
 * spacings,
 * zIndices,
});
 */
export const generateThemeFromTokens = ({
  borderWidths,
  colors,
  radii,
  fonts,
  fontSizes,
  fontWeights,
  lineHeights,
  boxShadows,
  spacings,
  zIndices,
}: GenerateThemeFromTokensArgs): GenericThemeShape => {
  // Default breakpoints 400px, 1024px, 1232px
  const breakpoints = ["25rem", "64rem", "77rem"];

  return {
    shadows: boxShadows,
    borderWidths,
    radii,
    breakpoints,
    colors,
    fonts,
    fontSizes,
    fontWeights,
    lineHeights,
    space: spacings,
    zIndices,
  };
};
