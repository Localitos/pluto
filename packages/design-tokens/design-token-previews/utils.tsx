import React from "react";
import map from "lodash/map";
import keys from "lodash/keys";
import camelCase from "lodash/camelCase";
import capitalize from "lodash/capitalize";
import invokeMap from "lodash/invokeMap";

import fontSizeTokens from "../src/tokens/font-size.tokens.json";
import fontWeightTokens from "../src/tokens/font-weight.tokens.json";
import colorTokens from "../src/tokens/color.tokens.json";
import iconSizeTokens from "../src/tokens/size.tokens.json";
import spaceTokens from "../src/tokens/space.tokens.json";
import borderStyleTokens from "../src/tokens/border-style.tokens.json";
import borderWidthTokens from "../src/tokens/border-width.tokens.json";
import borderRadiiTokens from "../src/tokens/border-radius.tokens.json";

import {
  createBorderRadiiPreview,
  createSpacePreview,
  createColorPreview,
  createFontSizePreview,
  createIconSizePreview,
  createFontWeightPreview,
  createBorderStylePreview,
  createBorderWidthPreview,
} from "./utils/previews";

import { hexToHsla, hexToRgb } from "./utils/colors";
import { Token, TokenEntry } from "./types/Token";
import { Column, Row } from "./types/TokenTable";

import { CopyToClipboardButton } from "./CopyToClipboardButton";

const getKey = (object: Record<string, unknown>): string => {
  return keys(object)[0];
};

const getValue = ([, token]: TokenEntry): string => {
  return token.value;
};

const getComment = ([, token]: TokenEntry): string => {
  return token.comment || "";
};

const formatTokenName = (prefix: string, suffix: string) => {
  const tokenName = camelCase(`${prefix}${capitalize(suffix)}`);
  return (
    <div style={{ display: "flex" }}>
      <div>{tokenName}</div>
      <div>
        <CopyToClipboardButton textToCopy={tokenName} />
      </div>
    </div>
  );
};

export const TOKENS = {
  [getKey(borderRadiiTokens)]: [
    { name: "Pixels", transform: getValue },
    { name: "Preview", transform: createBorderRadiiPreview },
  ],
  [getKey(borderWidthTokens)]: [
    { name: "Pixels", transform: getValue },
    { name: "Preview", transform: createBorderWidthPreview },
  ],
  [getKey(borderStyleTokens)]: [
    { name: "Style", transform: getValue },
    { name: "Preview", transform: createBorderStylePreview },
  ],
  [getKey(spaceTokens)]: [
    { name: "Pixels", transform: getComment },
    { name: "Rems", transform: getValue },
    { name: "Preview", transform: createSpacePreview },
  ],
  [getKey(iconSizeTokens)]: [
    { name: "Pixels", transform: getComment },
    { name: "Rems", transform: getValue },
    { name: "Preview", transform: createIconSizePreview },
  ],
  [getKey(fontWeightTokens)]: [
    { name: "Weight", transform: getValue },
    { name: "Preview", transform: createFontWeightPreview },
  ],
  [getKey(fontSizeTokens)]: [
    { name: "Pixels", transform: getComment },
    { name: "Rems", transform: getValue },
    { name: "Preview", transform: createFontSizePreview },
  ],
  [getKey(colorTokens)]: [
    { name: "Hex", transform: getValue },
    {
      name: "RGB",
      transform: ([, token]: TokenEntry): string => hexToRgb(token.value),
    },
    {
      name: "Hsla",
      transform: ([, token]: TokenEntry): string => hexToHsla(token.value),
    },
    { name: "Preview", transform: createColorPreview },
  ],
};

export const renderRows = (
  columns: Column[],
  tokens: Record<string, Record<string, Token>>
): Array<Row> => {
  const [tokenName, tokenEntries] = Object.entries(tokens)[0];

  return map(Object.entries(tokenEntries), (token) => {
    const [key] = token;

    const tokenRows: Row[] = invokeMap(columns, "transform", token) as Row[];

    return [formatTokenName(tokenName, key), ...tokenRows];
  }) as Row[];
};
