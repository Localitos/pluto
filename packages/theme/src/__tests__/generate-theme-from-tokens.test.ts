import { generateThemeFromTokens } from "../utils/generate-theme-from-tokens";

describe("generateThemeFromTokens", () => {
  it("should generate a theme in the desired shape based on design tokens provided", () => {
    expect(
      generateThemeFromTokens({
        borderWidths: { borderWidth0: "10px" },
        radii: { borderRadius10: "20px" },
        fonts: { fontFamilyModerat: "arial" },
        fontSizes: { fontSize30: "300px" },
        fontWeights: { fontWeightBold: "800" },
        lineHeights: { lineHeight30: "20px" },
        boxShadows: { shadow: "grey" },
        colors: { colorText: "black" },
        spacings: { space10: "28px" },
        zIndices: { zIndex0: "-20" },
      })
    ).toMatchSnapshot();
  });
});
