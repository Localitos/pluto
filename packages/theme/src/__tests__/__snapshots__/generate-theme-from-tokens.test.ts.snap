// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateThemeFromTokens should generate a theme in the desired shape based on design tokens provided 1`] = `
Object {
  "borderWidths": Object {
    "borderWidth0": "10px",
  },
  "breakpoints": Array [
    "25rem",
    "64rem",
    "77rem",
  ],
  "colors": Object {
    "colorText": "black",
  },
  "fontSizes": Object {
    "fontSize30": "300px",
  },
  "fontWeights": Object {
    "fontWeightBold": "800",
  },
  "fonts": Object {
    "fontFamilyModerat": "arial",
  },
  "lineHeights": Object {
    "lineHeight30": "20px",
  },
  "radii": Object {
    "borderRadius10": "20px",
  },
  "shadows": Object {
    "shadow": "grey",
  },
  "space": Object {
    "space10": "28px",
  },
  "zIndices": Object {
    "zIndex0": "-20",
  },
}
`;
